# Process the Datasets
This notebook takes as input the CodeNet and HumanEval-X datasets and creates the subsets that are used throughout the paper

In [1]:
import pandas as pd
from bs4 import BeautifulSoup
import numpy as np
import os
import re 

In [2]:
cwd = os.getcwd()

In [11]:
languages_humanevalx = ['Python', 'Java', 'C++', 'Go', 'Rust', 'JavaScript']
languages_codenet = ['Python', 'Java', 'C++', 'Go', 'Rust', 'JavaScript']

extension_map = {
    'Python' : 'py',
    'Java' : 'java',
    'C++' : 'cpp',
    'Go' : 'go',
    'JavaScript' : 'js',
    'Rust' : 'rs',
    'C#' : 'cs'
}

## HumanEval-X Subset

In [12]:
rust_df = pd.read_json(str(os.path.join(cwd, "datasets/codegeex/humaneval-x/rust/data/humaneval_rust.jsonl")), orient='records', lines=True)
rust_df['lang'] = 'Rust'

js_df = pd.read_json(os.path.join(cwd, "datasets/codegeex/humaneval-x/js/data/humaneval_js.jsonl"), orient='records', lines=True)
js_df['lang'] = 'JavaScript'

cpp_df = pd.read_json(os.path.join(cwd, "datasets/codegeex/humaneval-x/cpp/data/humaneval_cpp.jsonl"), orient='records', lines=True)
cpp_df['lang'] = 'C++'

go_df = pd.read_json(os.path.join(cwd, "datasets/codegeex/humaneval-x/go/data/humaneval_go.jsonl"), orient='records', lines=True)
go_df['lang'] = 'Go'

java_df = pd.read_json(os.path.join(cwd, "datasets/codegeex/humaneval-x/java/data/humaneval_java.jsonl"), orient='records', lines=True)
java_df['lang'] = 'Java'

pt_df = pd.read_json(os.path.join(cwd, "datasets/codegeex/humaneval-x/python/data/humaneval_python.jsonl"), orient='records', lines=True)
pt_df['lang'] = 'Python'

In [13]:
humanevalx = pd.concat([rust_df, js_df, cpp_df, go_df, java_df, pt_df])
humanevalx['id'] = humanevalx['task_id'].apply(lambda x : x.split('/')[1])

In [14]:
humanevalx.head()

task_id                                             prompt  \
0  Rust/0  \n/*\n Check if in given list of numbers, are ...   
1  Rust/1  \n/*\n Input to this function is a string cont...   
2  Rust/2  \n/*\n Given a positive floating point number,...   
3  Rust/3  \n/*\n You're given a list of deposit and with...   
4  Rust/4  \n/*\n For a given list of input numbers, calc...   

                                         declaration  \
0  \nuse std::{slice::Iter, cmp::{max, self}, mem...   
1  \nuse std::{slice::Iter, cmp::{max, self}, mem...   
2  \nuse std::{slice::Iter, cmp::{max, self}, mem...   
3  \nuse std::{slice::Iter, cmp::{max, self}, mem...   
4  \nuse std::{slice::Iter, cmp::{max, self}, mem...   

                                  canonical_solution  \
0  \n    for i in 0..numbers.len(){\n        for ...   
1  \n    let mut result:Vec<String> = vec![];\n  ...   
2                    \n    return number % 1.0;\n}\n   
3  \n\nlet mut balance:i32 = 0;\nfor op in operat...   
4  \n    let mean:f32 = numbers.iter().fold(0.0,|...   

                                                test example_test  lang  \
0  \n#[cfg(test)]\nmod tests {\n    use super::*;...         None  Rust   
1  \n#[cfg(test)]\nmod tests {\n    use super::*;...         None  Rust   
2  \n#[cfg(test)]\nmod tests {\n    use super::*;...         None  Rust   
3  \n#[cfg(test)]\nmod tests {\n    use super::*;...         None  Rust   
4  \n#[cfg(test)]\nmod tests {\n    use super::*;...         None  Rust   

  import docstring test_setup text id  
0    NaN       NaN        NaN  NaN  0  
1    NaN       NaN        NaN  NaN  1  
2    NaN       NaN        NaN  NaN  2  
3    NaN       NaN        NaN  NaN  3  
4    NaN       NaN        NaN  NaN  4

In [15]:
len(humanevalx.groupby('task_id'))

984

In [16]:
translation_pairs = []

for task_id, group in humanevalx.groupby('id'):
    
    for source in languages_humanevalx:
        for target in languages_humanevalx:
            obj = {}
            obj['id'] = task_id
            
            if source != target:
                obj['source_lang'] = source
                mask = group['lang'].str.lower() == source.lower()
                input_row = group[mask].iloc[0]
                obj['input_code'] = input_row['declaration'] + input_row['canonical_solution']
                obj['target_lang'] = target
                mask = group['lang'].str.lower() == target.lower()
                target_row = group[mask].iloc[0]
                obj['ground_truth'] = target_row['declaration'] + target_row['canonical_solution']
                obj['target_signature'] = target_row['declaration']
                obj['test_code'] = target_row['test']
                
                translation_pairs.append(obj)
                
translation_df = pd.DataFrame(translation_pairs)

In [17]:
translation_df.shape

(4920, 7)

In [18]:
translation_df.head()

id source_lang                                         input_code  \
0  0      Python  from typing import List\n\n\ndef has_close_ele...   
1  0      Python  from typing import List\n\n\ndef has_close_ele...   
2  0      Python  from typing import List\n\n\ndef has_close_ele...   
3  0      Python  from typing import List\n\n\ndef has_close_ele...   
4  0      Python  from typing import List\n\n\ndef has_close_ele...   

  target_lang                                       ground_truth  \
0        Java  import java.util.*;\nimport java.lang.*;\n\ncl...   
1         C++  #include<stdio.h>\n#include<vector>\n#include<...   
2          Go  \nfunc HasCloseElements(numbers []float64, thr...   
3        Rust  \nuse std::{slice::Iter, cmp::{max, self}, mem...   
4  JavaScript  \nconst hasCloseElements = (numbers, threshold...   

                                    target_signature  \
0  import java.util.*;\nimport java.lang.*;\n\ncl...   
1  #include<stdio.h>\n#include<vector>\n#include<...   
2  \nfunc HasCloseElements(numbers []float64, thr...   
3  \nuse std::{slice::Iter, cmp::{max, self}, mem...   
4  \nconst hasCloseElements = (numbers, threshold...   

                                           test_code  
0  public class Main {\n    public static void ma...  
1  #undef NDEBUG\n#include<assert.h>\nint main(){...  
2  func TestHasCloseElements(t *testing.T) {\n   ...  
3  \n#[cfg(test)]\nmod tests {\n    use super::*;...  
4  const testHasCloseElements = () => {\n  consol...

#### 888 is the minimum sample size for a population of 4920 translation pairs with 99.9% CI

In [19]:
stratified = translation_df.groupby(['source_lang', 'target_lang'])

dfs = []

for name, group in stratified:
    dfs.append(group.sample(n=35, random_state=1))
    
humaneval_x_pairs_subset = pd.concat(dfs).reset_index(drop=True)

In [20]:
humaneval_x_pairs_subset.shape

(1050, 7)

In [21]:
humaneval_x_pairs_subset.to_json('./datasets/humanevalx_dataset_subset.jsonl', orient='records', lines=True)

In [22]:
translation_df.to_json('./datasets/humanevalx_dataset_all.jsonl', orient='records', lines=True)

## CodeNet

In [ ]:
problem_descriptions_path = './datasets/Project_CodeNet/problem_descriptions'

In [ ]:
import os
import random

# Set the seed for reproducibility
seed = 1
random.seed(seed)

# Get a list of all files in the directory
all_files = [os.path.join(problem_descriptions_path, f) for f in os.listdir(problem_descriptions_path) if os.path.isfile(os.path.join(problem_descriptions_path, f))]

print(len(all_files))

selected_files = all_files

In [ ]:
# Function to extract the fuzzy tests from HTML content
def extract_input_output(file_path):
    with open(file_path, 'r', encoding='utf-8') as file:
        content = file.read()
    soup = BeautifulSoup(content, 'html.parser')
    
    # Extract input
    input_tag = soup.find('h3', string='Sample Input 1')
    if input_tag:
        input_value = input_tag.find_next('pre').get_text(strip=True)
    else:
        input_value = None
        
    input_tag = soup.find('h3', string='Sample Input 2')
    if input_tag:
        input_value_2 = input_tag.find_next('pre').get_text(strip=True)
    else:
        input_value_2 = None
        
    input_tag = soup.find('h3', string='Sample Input 3')
    if input_tag:
        input_value_3 = input_tag.find_next('pre').get_text(strip=True)
    else:
        input_value_3 = None
    
    # Extract output
    output_tag = soup.find('h3', string='Sample Output 1')
    if output_tag:
        output_value = output_tag.find_next('pre').get_text(strip=True)
    else:
        output_value = None
        
    # Extract output
    output_tag = soup.find('h3', string='Sample Output 2')
    if output_tag:
        output_value_2 = output_tag.find_next('pre').get_text(strip=True)
    else:
        output_value_2 = None
        
    # Extract output
    output_tag = soup.find('h3', string='Sample Output 3')
    if output_tag:
        output_value_3 = output_tag.find_next('pre').get_text(strip=True)
    else:
        output_value_3 = None
    
    return input_value, output_value, input_value_2, output_value_2, input_value_3, output_value_3

# Dictionary to store the (input, output) pairs
input_output_dict = {}

# Process each selected file
for file_path in selected_files:
    input_value, output_value, input_value_2, output_value_2, input_value_3, output_value_3 = extract_input_output(file_path)
    filename = filename = os.path.basename(file_path).split('.')[0]
    if input_value is not None and input_value_2 is not None and output_value is not None and output_value_2 is not None and input_value_3 is not None and output_value_3 is not None:
        input_output_dict[filename] = (input_value, output_value, input_value_2, output_value_2, input_value_3, output_value_3)

In [ ]:
#Number of problems with three input-output fuzzy tests
len(input_output_dict)

In [ ]:
#Given that CodeNet works through stdin. We need to make sure there is a newline so the data in stdin gets passed to the function's reader
def add_newline_if_missing(s):
    if not s.endswith('\n'):
        s += '\n'
    return s

In [ ]:
entries = []

for problemid, inputoutput in input_output_dict.items():    
    #We read the metadata for this problem
    metadata_path = f'./datasets/Project_CodeNet/metadata/{problemid}.csv'
    df = pd.read_csv(metadata_path)
    df = df[(df.status == 'Accepted')]
    skipit = False
    local_entries = []
    
    if df.empty == True:
        continue
    
    for lang in languages_codenet:
        obj = {}

        obj['problem'] = problemid
        obj['stdin_input_1'] = add_newline_if_missing(inputoutput[0])
        obj['expected_output_1'] = inputoutput[1]
        obj['stdin_input_2'] = add_newline_if_missing(inputoutput[2])
        obj['expected_output_2'] = inputoutput[3]
        obj['stdin_input_3'] = add_newline_if_missing(inputoutput[4])
        obj['expected_output_3'] = inputoutput[5]
        dflang = df[(df.language == lang)]
        
        #We keep programs that are less than 1kb (as a proxy to token length. This is to ensure we don't have problems with context length)
        dflang = dflang[(dflang.code_size < 1024)]
        
        if dflang.empty:
            #No parallel data in this language. Moving on to the next problem
            skipit = True
            break
        
        #Sample one random solutions to this problem, from all the solutions
        row = dflang.sample(n=1, random_state=1).iloc[0]
        submission_id = row['submission_id']
        obj['size_bytes'] = row['code_size']
        
        #Get the content from the submission
        with open(f'./datasets/Project_CodeNet/data/{problemid}/{lang}/{submission_id}.{extension_map[lang]}') as f:
            content = f.read()
            obj['language'] = lang
            obj['code'] = content
        
        local_entries.append(obj)
        
    if not skipit:
        entries.extend(local_entries)

In [ ]:
codenet_subset = pd.DataFrame(entries)

In [ ]:
codenet_subset.shape

In [ ]:
codenet_subset.head()

In [ ]:
codenet_translation_pairs = []

#Build the possible translation pairs using all the problems found previously
for problem, group in codenet_subset.groupby('problem'):
    
    for source in languages_codenet:
        for target in languages_codenet:
            obj = {}
            obj['task_id'] = problem
            
            if source != target:
                obj['source_lang'] = source
                mask = group['language'].str.lower() == source.lower()
                input_row = group[mask].iloc[0]
                obj['input_code'] = input_row['code']
                obj['target_lang'] = target
                mask = group['language'].str.lower() == target.lower()
                target_row = group[mask].iloc[0]
                obj['ground_truth'] = target_row['code']
                
                obj['stdin_input_1'] = input_row['stdin_input_1']
                obj['expected_output_1'] = input_row['expected_output_1']
                obj['stdin_input_2'] = input_row['stdin_input_2']
                obj['expected_output_2'] = input_row['expected_output_2']
                obj['stdin_input_3'] = input_row['stdin_input_3']
                obj['expected_output_3'] = input_row['expected_output_3']
                
                #Sanity check for alignment
                assert target_row['expected_output_1'] == input_row['expected_output_1']
                assert target_row['expected_output_2'] == input_row['expected_output_2']
                assert target_row['expected_output_3'] == input_row['expected_output_3']
                      
                codenet_translation_pairs.append(obj)
                
codenet_translation_df = pd.DataFrame(codenet_translation_pairs)

In [ ]:
codenet_translation_df.shape

#### We need at least 1013 samples from 15660 translation pairs for 99.9% CI

In [ ]:
codenet_stratified = codenet_translation_df.groupby(['source_lang', 'target_lang'])

codenet_dfs = []

for name, group in codenet_stratified:
    codenet_dfs.append(group.sample(n=35, random_state=1))
    
codenet_pairs_subset = pd.concat(codenet_dfs).reset_index(drop=True)

In [ ]:
codenet_pairs_subset.shape

In [ ]:
codenet_pairs_subset.head()

In [ ]:
codenet_translation_df.to_json('./datasets/codenet_pairs_all.jsonl', orient='records', lines=True)

In [ ]:
codenet_pairs_subset.to_json('./datasets/codenet_pairs_subset.jsonl', orient='records', lines=True)

## TransCoder Test Set (curated)

In [5]:
unitrans_df = pd.read_json('./datasets/FSE-24-UniTrans/cleaned_data/testable_samples.jsonl', orient='records', lines=True)

In [6]:
unitrans_df.head()

id  \
0                            ADD_1_TO_A_GIVEN_NUMBER   
1                          ADD_1_TO_A_GIVEN_NUMBER_1   
2   ANALYSIS_OF_ALGORITHMS_SET_2_ASYMPTOTIC_ANALYSIS   
3  AREA_OF_THE_CIRCLE_THAT_HAS_A_SQUARE_AND_A_CIR...   
4                   AREA_SQUARE_CIRCUMSCRIBED_CIRCLE   

                                                 cpp  \
0  int addOne ( int x ) {\n  int m = 1;\n  while ...   
1  int addOne ( int x ) {\n  return ( - ( ~ x ) )...   
2  int search ( int arr [ ], int n, int x ) {\n  ...   
3  float getArea ( int a ) {\n  float area = ( M_...   
4  int find_Area ( int r ) {\n  return ( 2 * r * ...   

                                                java  \
0  static int addOne ( int x ) {\n  int m = 1 ;\n...   
1  static int addOne ( int x ) {\n  return ( - ( ...   
2  static int search ( int arr [ ] , int n , int ...   
3  static float getArea ( int a ) {\n  float area...   
4  static int find_Area ( int r ) {\n  return ( 2...   

                                              python  
0  def addOne ( x ) :\n    m = 1\n    while ( x &...  
1       def addOne ( x ) :\n    return ( - ( ~ x ) )  
2  def search ( arr , n , x ) :\n    i = 0\n    f...  
3  def getArea ( a ) :\n    area = ( math.pi * a ...  
4    def find_Area ( r ) :\n    return ( 2 * r * r )

In [21]:
unitrans_df.shape

(568, 4)

In [7]:
df_melted = pd.melt(unitrans_df, id_vars=['id'], var_name='language', value_name='value')

In [8]:
df_melted.head()

id language  \
0                            ADD_1_TO_A_GIVEN_NUMBER      cpp   
1                          ADD_1_TO_A_GIVEN_NUMBER_1      cpp   
2   ANALYSIS_OF_ALGORITHMS_SET_2_ASYMPTOTIC_ANALYSIS      cpp   
3  AREA_OF_THE_CIRCLE_THAT_HAS_A_SQUARE_AND_A_CIR...      cpp   
4                   AREA_SQUARE_CIRCUMSCRIBED_CIRCLE      cpp   

                                               value  
0  int addOne ( int x ) {\n  int m = 1;\n  while ...  
1  int addOne ( int x ) {\n  return ( - ( ~ x ) )...  
2  int search ( int arr [ ], int n, int x ) {\n  ...  
3  float getArea ( int a ) {\n  float area = ( M_...  
4  int find_Area ( int r ) {\n  return ( 2 * r * ...

In [9]:
df_melted["id"].nunique()

568

In [10]:
import os
import random

translation_pairs = []
languages = ["python", "java", "cpp"]

map_unitrans = {
    "python" : "Python",
    "java" : "Java",
    "cpp" : "C++"
}

# Set the seed for reproducibility
seed = 1
random.seed(seed)

results = []

for name, group in df_melted.groupby("id"):
    group_entries = []
    python_found = False
    cpp_found = False
    java_found = False


    for index, row in group.iterrows():
        problem = row['id']
        cpp_path = f"./datasets/FSE-24-UniTrans/cleaned_data/transcoder_evaluation_gfg/cpp/{problem}.cpp"
        java_path = f"./datasets/FSE-24-UniTrans/cleaned_data/transcoder_evaluation_gfg/java/{problem}.java"
        python_path = f"./datasets/FSE-24-UniTrans/cleaned_data/transcoder_evaluation_gfg/python/{problem}.py"

        obj = {}

        source = row['language']
        
        for target in languages:
            obj = {}
            obj['id'] = problem
            
            if source != target:

                try:
                    if target == "cpp":
                        with open(cpp_path, "r") as f:
                            obj["test_code"] = f.read()
                    if target == "java":
                        with open(java_path, "r") as f:
                            obj["test_code"] = f.read()
                    if target == "python":
                        with open(python_path, "r") as f:
                            obj["test_code"] = f.read()

                    obj["source_lang"] = map_unitrans[source]
                    obj['input_code'] = row['value']
                    obj['target_lang'] = map_unitrans[target]
                    mask = group['language'].str.lower() == target.lower()
                    target_row = group[mask].iloc[0]
                    obj['ground_truth'] = target_row['value']
                    obj['target_signature'] = "f_filled"

                    translation_pairs.append(obj)

                except Exception as e:
                    #Parallel data is missing for this problem. This is expected
                    continue

In [11]:
translation_df = pd.DataFrame(translation_pairs)

In [17]:
translation_df.shape

(2826, 7)

In [18]:
len(translation_df.groupby("id"))

568

In [19]:
stratified = translation_df.groupby(['source_lang', 'target_lang'])

dfs = []

for name, group in stratified:
    dfs.append(group.sample(n=175, random_state=1))
    
unitrans_pairs_subset = pd.concat(dfs)

In [20]:
unitrans_pairs_subset.head()

id  \
186          CHECK_LARGE_NUMBER_DIVISIBLE_11_NOT   
1457               LONGEST_REPEATING_SUBSEQUENCE   
1087   FIND_SUM_UNIQUE_SUB_ARRAY_SUM_GIVEN_ARRAY   
631                   C_PROGRAM_FACTORIAL_NUMBER   
515   COUNT_PALINDROMIC_SUBSEQUENCE_GIVEN_STRING   

                                              test_code source_lang  \
186   // Copyright (c) 2019-present, Facebook, Inc.\...         C++   
1457  // Copyright (c) 2019-present, Facebook, Inc.\...         C++   
1087  // Copyright (c) 2019-present, Facebook, Inc.\...         C++   
631   // Copyright (c) 2019-present, Facebook, Inc.\...         C++   
515   // Copyright (c) 2019-present, Facebook, Inc.\...         C++   

                                             input_code target_lang  \
186   int check ( string str ) {\n  int n = str . le...        Java   
1457  int findLongestRepeatingSubSeq ( string str ) ...        Java   
1087  long long int findSubarraySum ( int arr [ ], i...        Java   
631   unsigned int factorial ( unsigned int n ) {\n ...        Java   
515   int countPS(string str)\n{\n  int N = str.leng...        Java   

                                           ground_truth target_signature  
186   static boolean check ( String str ) {\n  int n...         f_filled  
1457  static int findLongestRepeatingSubSeq ( String...         f_filled  
1087  static int findSubarraySum ( int [ ] arr , int...         f_filled  
631   static int factorial ( int n ) {\n  if ( n == ...         f_filled  
515   static int countPS ( String str ) {\n  int N =...         f_filled

In [ ]:
unitrans_pairs_subset.shape

In [ ]:
unitrans_pairs_subset.to_json('./datasets/unitrans_dataset_subset.jsonl', orient='records', lines=True)
translation_df.to_json('./datasets/unitrans_dataset_all.jsonl', orient='records', lines=True)